modules/core/arrow-annotations-processor/src/main/java/arrow/generic/ProductProcessor.kt

```diff
@@ -0,0 +1,84 @@
+package arrow.generic
+
+import com.google.auto.service.AutoService
+import arrow.common.utils.AbstractProcessor
+import arrow.common.utils.asClassOrPackageDataWrapper
+import arrow.common.utils.knownError
+import me.eugeniomarletti.kotlin.metadata.KotlinClassMetadata
+import me.eugeniomarletti.kotlin.metadata.extractFullName
+import me.eugeniomarletti.kotlin.metadata.isDataClass
+import me.eugeniomarletti.kotlin.metadata.kotlinMetadata
+import me.eugeniomarletti.kotlin.metadata.proto
+import org.jetbrains.kotlin.serialization.ProtoBuf
+import java.io.File
+import javax.annotation.processing.Processor
+import javax.annotation.processing.RoundEnvironment
+import javax.lang.model.SourceVersion
+import javax.lang.model.element.TypeElement
+
+import javax.lang.model.element.Element
+
+@AutoService(Processor::class)
+class ProductProcessor : AbstractProcessor() {
+
+    private val annotatedProduct = mutableListOf<AnnotatedGeneric>()
+
+    override fun getSupportedSourceVersion(): SourceVersion = SourceVersion.latestSupported()
+
+    override fun getSupportedAnnotationTypes() = setOf(productAnnotationClass.canonicalName)
+
+    override fun onProcess(annotations: Set<TypeElement>, roundEnv: RoundEnvironment) {
+        annotatedProduct += roundEnv
+                .getElementsAnnotatedWith(productAnnotationClass)
+                .map(this::evalAnnotatedProductElement)
+
+        if (roundEnv.processingOver()) {
+            val generatedDir = File(this.generatedDir!!, "").also { it.mkdirs() }
+            ProductFileGenerator(annotatedProduct, generatedDir).generate()
```

Comment:
I think it'd be good to generate each file under the target class package, like other processors do (Dagger, ButterKnife). Those do it bc they need package protected visibility to access injectable fields on targets, which is not needed here, but that's still handy for traceability. The need to browse a generated class is usual and it's good to have them right under the same package structure of the target class if you navigate through the `build/` dir on your project.

Reply:
I agree, if we want to do that we should create an issue and tackle it in a different PR because that is not just specific to the `ProductProcessor` or Generic but to all other processors as well for all other annotations in Arrow.

Reply:
Agree. Created #765